### PR TITLE
feat(PPDSC-2120): popover v1

### PR DESCRIPTION
**Pull request by mstuartf**
[PPDSC-2120](https://nidigitalsolutions.jira.com/browse/PPDSC-2120)

**What**

1. Background: add Popover component
2. Created BaseFloatingElement component to be used by Tooltip and Popover
3. What does the reviewers should expect

**I have done:**
- [x] Written unit tests against changes
- [ ] Written functional tests against the component and/or NewsKit site
- [ ] Updated relevant documentation

**I have tested manually:**
- [x] The feature's functionality is working as expected on Chrome, Firefox, Safari and Edge
- [x] The screen reader reads and flows through the elements as expected.
- [x] There are no new errors in the browser console coming from this PR.
- [ ] When visual test is not added, it renders correctly on different browsers and mobile viewports (Safari, Firefox, small mobile viewport, tablet)
- [ ] The Playground feature is working as expected

Differences to ticket:
* Removed support for `defaultOpen` prop for Tooltip and Popover due to issue with 3rd party lib
* Added `aria-labelledby` to Popover floating element to prevent `aria-dialog-name` issue (also requires giving the reference element an id if it does not already have one).
* `overrides.distance` must be a px string
* Added a `pointer.padding` default / override so that the pointer does not reach the edge of the element (this looks strange with border radius applied). 
* Changed `fallbackPlacement` to `fallbackBehaviour` for clarity.

### CODE REVIEW

**Comment by newskit-engineering** [comment]
You can preview these changes on: 
  Documentation Site: http://ncu-newskit-docs.s3-website-eu-west-1.amazonaws.com/ppdsc-2120-popover-v1 
  Storybook: http://ncu-newskit-docs.s3-website-eu-west-1.amazonaws.com/ppdsc-2120-popover-v1/storybook

File: src/base-floating-element/utils.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  const rawValue = getRawValue(path, props, overridePath, defaultsObjectKey);
+  const parsedValue = rawValue ? parseRawValue(rawValue) : null;
+  return parsedValue || undefined;
+};
+
+export const showOverridePxWarnings = (
+  parsedValue: number | undefined,
+  overridePath: string,
+): void => {
+  if (!parsedValue) {
```
Comment (mutebg) [inline src/base-floating-element/utils.ts:58]: On few places we have the fallowing: 
```suggestion
  if (process.env.NODE_ENV !== 'production' && !parsedValue) {
```

File: src/base-floating-element/utils.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+// They only accept strings and a console warning will display if the value passed
+// is not a valid token or px string.
+// This function fetches the override value and returns the px value as an integer.
+// If override is not a valid token or px string, this function returns undefined.
+export const getOverridePxValue = (
+  path: string,
+  props: ThemeProp & {overrides?: unknown},
+  overridePath: string,
+  defaultsObjectKey: string,
+): number | undefined => {
```
Comment (mutebg) [inline src/base-floating-element/utils.ts:48]: I believe you can simplify this and getRawValue using something like:
```ts
// getToken will return you the tokenName ( space030 ) or px value like 10px
const token = getToken(
    {theme, overrides},
    `${path}.pointer`,
    'pointer.padding',
    'padding',
  );
// check if that value exist in spacePreset as key or use the raw value.  
 const rawValue = theme['spacePresets'][token] || token;
 ```

File: src/base-floating-element/base-floating-element.tsx [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    {theme, overrides},
+    'distance',
+    'distance',
+  );
+  const pointerPadding = getOverridePxValue(
+    `${path}.pointer`,
+    {theme, overrides},
+    'pointer.padding',
+    'padding',
+  );
```
Comment (mutebg) [inline src/base-floating-element/base-floating-element.tsx:55]: I am not sure how this padding is used, I made a story with the same component with 2 different padding overrides and they look exactly the same?
Is this a really a padding or some kind of space between elements?

**Comment by mutebg** [comment]
few more things from the ticket:
- its mention that Esc key shout close the popover
- In the next ticket, we will need to add the Header, and Close button props, wondering if we should implement the Content component now since is part of that ticket, or leave it for later when we have the header and close btn

File: src/popover/__tests__/popover.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+StoryPopoverPositionsVisualTest.storyName = 'popover-positions-visual-test';
+
+export const StoryPopoverBehavioursVisualTest = () => (
+  <>
+    <PopoverWithBoundary visualTest fallbackBehaviour={['flip']} />
+    <PopoverWithBoundary visualTest fallbackBehaviour={['shift']} />
+    <PopoverWithBoundary visualTest fallbackBehaviour={['flip', 'shift']} />
+  </>
+);
+StoryPopoverBehavioursVisualTest.storyName = 'popover-behaviours-visual-test';
```
[the remark this answers is not in the document]
Comment (mstuartf) [reply src/popover/__tests__/popover.stories.tsx:273]: Oh nice, thanks.

File: src/tooltip/tooltip.tsx [hunk 5]
```diff
@@ -1,137 +1,72 @@
...
 
-  const childrenProps = {
-    ...labelOrDescProps,
-    ...children.props,
+    return attrs;
   };
 
-  if (!content) {
-    return children;
-  }
```
[the remark this answers is not in the document]
Comment (mstuartf) [reply src/tooltip/tooltip.tsx:92]: The BaseFloatingElement now has this logic. Do you mean that we shouldn't have it for the Popover?

File: src/tooltip/types.ts [hunk 6]
```diff
@@ -1,35 +1,8 @@
...
 
-export interface TooltipProps
-  extends Omit<React.HTMLAttributes<HTMLDivElement>, 'title' | 'defaultValue'> {
-  children: React.ReactElement & {
-    ref?: React.Ref<unknown>;
-  };
-  content: React.ReactNode;
-  open?: boolean;
-  defaultOpen?: boolean;
+export type TooltipProps = Omit<FloatingElementProps, 'trigger'> & {
```
[the remark this answers is not in the document]
Comment (mstuartf) [reply src/tooltip/types.ts:5]: I'll check with Mike M

File: src/base-floating-element/base-floating-element.tsx (hunk 3 above)
Comment (mstuartf) [reply src/base-floating-element/base-floating-element.tsx:55]: This is passed to the `arrow` middleware: https://floating-ui.com/docs/arrow#padding.

I used `padding` to mirror their API. But (clearly) it's a confusing name. Do you have any suggestions?

File: src/tooltip/tooltip.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (mutebg) [reply src/tooltip/tooltip.tsx:92]: now I see, if we already have it at Base component we might remove it from Popover

File: src/base-floating-element/base-floating-element.tsx (hunk 3 above)
Comment (mutebg) [reply src/base-floating-element/base-floating-element.tsx:55]: something like `offset` or `edgeOffset` maybe?

**Comment by mutebg** [comment]
> > * its mention that Esc key shout close the popover
> 
> I clarified this with Mike M and he said we should not do this. See comments on the ticket: https://nidigitalsolutions.jira.com/browse/PPDSC-2120
> 
> > * In the next ticket, we will need to add the Header, and Close button props, wondering if we should implement the Content component now since is part of that ticket, or leave it for later when we have the header and close btn
> 
> I'd rather not add any more to this PR if that's ok.
> 
> @mutebg

Thanks, I should have read the comments as well. Seems good

File: src/tooltip/types.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mstuartf) [reply src/tooltip/types.ts:5]: Omitting `onDismiss` and `restoreFocusTo` as these are not applicable to the Tooltip but keeping `fallbackBehaviour`.

File: src/base-floating-element/base-floating-element.tsx (hunk 3 above)
Comment (mstuartf) [reply src/base-floating-element/base-floating-element.tsx:55]: `edgeOffset` is good.

File: src/popover/__tests__/popover.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+      <StorybookSubHeading>String content</StorybookSubHeading>
+      <StyledContainer style={getPlacementPosition('top')}>
+        <PopoverWithBtn content="Popover content" />
+      </StyledContainer>
+    </div>
+    <div>
+      <StorybookSubHeading>HTML content</StorybookSubHeading>
+      <StyledContainer style={getPlacementPosition('top')}>
+        <PopoverWithBtn content={<Button size="small">Click me</Button>} />
+      </StyledContainer>
```
[the remark this answers is not in the document]
Comment (mstuartf) [reply src/popover/__tests__/popover.stories.tsx:87]: Well spotted. I think we had copied their tutorial CSS which had `pointer-events: none`. Removed.

File: src/theme/__tests__/__snapshots__/theme.test.ts.snap [hunk 8]
```diff
@@ -1976,6 +1976,22 @@ Object {
...
+  "popover": Object {
+    "distance": "space050",
+    "panel": Object {
+      "paddingBlock": "space040",
+      "paddingInline": "space040",
+      "stylePreset": "popoverPanel",
+      "typographyPreset": "utilityLabel010",
+    },
+    "pointer": Object {
+      "padding": "space030",
```
Comment (mutebg) [inline src/theme/__tests__/__snapshots__/theme.test.ts.snap:1988]: I guess this snapshot need updating

**Review by mutebg — COMMENTED** [review]
Everything seems good to me, ready to approve it. just let me know when you fix your snapshots!

File: src/theme/__tests__/__snapshots__/theme.test.ts.snap (hunk 8 above)
Comment (mstuartf) [reply src/theme/__tests__/__snapshots__/theme.test.ts.snap:1988]: 🤦

**Review by mutebg — APPROVED** [review]

**Review by mutebg — APPROVED** [review]

**Pull request merged**